pkg/resource/reference_resolver.go

```diff
@@ -0,0 +1,140 @@
+/*
+Copyright 2019 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resource
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	corev1 "k8s.io/api/core/v1"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// ReferenceResolver is an interface for resolving cross resource attribute references.
+// It should be implemented by strcuts that are used for custom resources feilds.
+// See #874 for more information
+type ReferenceResolver interface {
+	// Resolve retrieve and builds the referenced attribute
+	Resolve(reader client.Reader) (value interface{}, err error)
+
+	// Assign accepts a resource object, and assigns the given value to the corresponding property
+	Assign(resource interface{}, value interface{}) error
+}
```

Comment:
For instance, the VPCID referencer will be implemented as:

```go
type VPCIDResolver struct {
	VPC *LocalObjectReference
}

func (v *VPCIDResolver) Resolve(reader client.Reader) (interface{}, error) {

	var vpc network.VPC
	if err := reader.Get(ctx, v.VPC, &vpc); err != nil {
		return nil, err
	}

	return vpc.Id, nil
}

func (v *VPCIDResolver) Assign(resource interface{}, value interface{}) error {
	subnet, isExpectedType := resource.(*network.Subnet)
	if !isExpectedType {
		return fmt.Errorf("unexpected resource type")
	}

	subnet.Spec.VPCID = value.(string)
	return nil
}
```

and then will be used in [Subnet type](https://github.com/crossplaneio/stack-aws/blob/master/apis/network/v1alpha2/subnet_types.go#L28) like:
```go
VPCIDRef *VPCIDResolver `json:"vpcIdRef,omitempty" managed_resource:"reference_resolver"`
```

Reply:
the `managed_resource:"reference_resolver"` is intended to help readability of the code, but visually linking it to managed reconciler's reference resolution mechanism.

Reply:
I'm still concerned that this results in a an API like:

```yaml
vpcIDRef:  # This looks like it will be an object reference
  vpc:  # ...but it is not
    name: myVPC
```

Reply:
@negz for single value object resolvers like `VPCIDResolver`, we could embed the reference:
```
type VPCIDResolver struct {
	*LocalObjectReference
}
```
But for composite attributes, I don't see any other way around this.

Reply:
https://github.com/crossplaneio/crossplane-runtime/issues/49

> for single value object resolvers like VPCIDResolver, we could embed the reference:

I'd prefer this over the current pattern. I think implementing our own type would be better, for the reasons outlined in the above issue. Using our own type would let us make `name` a required field.

> But for composite attributes, I don't see any other way around this.

I'm still not convinced this use case exists. Did we have any examples of composite attributes in which the additional attributes could not be derived from a single attribute? i.e. Deriving the project ID from the network name.

Reply:
`Subnetwork` in GCP is the only case that I can think of. If we can retrieve the ProjectId from the network, we are good Otherwise we will need some api object from which we can retrieve the project id

Reply:
> If we can retrieve the ProjectId from the network, we are good 

I'm 99.% sure we can. We'd follow the network's provider ref to its provider resource and thus the project used by that provider.

Reply:
> We'd follow the network's provider ref to its provider resource and thus the project used by that provider.

It's worth noting that for this case whatever is doing the resolving would need a `client.Reader` in order to get and inspect the providerRef similar to the example I posted at https://github.com/crossplaneio/crossplane-runtime/pull/46#discussion_r334722588. I didn't quite grok how automatic the current implementation is (in that the resolvers in the current implementation never actually talk to Kubernetes) when writing this comment initially.

My preference is still for the simpler API - i.e. `name` only. In the case of the subnetwork->network relationship it would be redundant to require subnet authors to specify which provider the network was in given that:

1. We can discover that.
1. There's only one valid provider - the one the network references.

Reply:
I agree. I'm planning to only use a name, by embedding a `LocalObjectReference` in the type

```go
type VPCIDResolver struct {
	*LocalObjectReference
}
```

This would resolve in a yaml like
```yaml
spec:
  ...
  vpcIdResolver:
    name: <vpcname>
  ...
```